### PR TITLE
Add gitlab as source

### PR DESCRIPTION
**Pull request by nellshamrell**

### CODE REVIEW

**Comment by nellshamrell** [comment]
Adds gitlab as outlined in #850

File: routes/originGitLab.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+router.get(
+  '/:namespace/:project/:revisions',
+  asyncMiddleware(async (request, response) => {
+    try {
+
+      const { namespace, project } = request.params
+
+      const project_info = await gitlab.Projects.search(project)
+      let project_match = getExactProjectMatch(namespace, project, project_info)
```
Comment (MichaelTsengLZ) [inline routes/originGitLab.js:22]: Nit: const

File: routes/originGitLab.js [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+      const tags = await gitlab.Tags.all(project_match.id)
+
+      let unsorted = tags.map(tag => {
```
Comment (MichaelTsengLZ) [inline routes/originGitLab.js:26]: Nit: const

File: routes/originGitLab.js [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+  return project_names
+}
+
+async function getUserProjects(username) {
+  const user_response = await getUsers(username)
+
+  const user_projects = await gitlab.Users.projects(user_response[0].id)
+
+  let user_project_names = user_projects.map(project => {
```
Comment (MichaelTsengLZ) [inline routes/originGitLab.js:111]: Nit: const

File: routes/originGitLab.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+// We can't filter on namespace when searching for a project
+// by name through the GitLab API, it will return all projects
+// with matches for the search text in their project name
+// This filters them by namespace
+function getExactProjectMatch(namespace, project, projects) {
+  let exact_match
+
+  projects.filter((item) => {
```
Comment (MichaelTsengLZ) [inline routes/originGitLab.js:85]: Nit: this works. But normally I will use [find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find)

File: app.js [hunk 5]
```diff
@@ -141,6 +141,7 @@ function createApp(config) {
   app.use('/origins/pypi', require('./routes/originPyPi')())
   app.use('/origins/rubygems', require('./routes/originRubyGems')())
   app.use('/origins/deb', require('./routes/originDeb')())
+  app.use('/origins/gitlab', require('./routes/originGitLab')())
```
Comment (MichaelTsengLZ) [inline app.js:144]: Is this route for search in the UI?

**Review by MichaelTsengLZ — APPROVED** [review]
I'm not that familiar with the process of adding a new component type.

File: app.js (hunk 5 above)
Comment (nellshamrell) [reply app.js:144]: Yes indeed!

**Pull request merged**